tests/Microsoft.Identity.Test.Unit.net45/Integration/DeviceCodeFlowIntegrationTest.cs

```diff
@@ -0,0 +1,169 @@
+        }
+
+
+        [TestMethod]
+        [Timeout(60 * 1000)] // 1 min timeout
+        public async Task DeviceCodeFlowTestAsync()
+        {
+            Debug.WriteLine("Fetching user from lab");
+            LabResponse labResponse = LabUserHelper.GetDefaultUser();
+
+            Debug.WriteLine("Calling AcquireTokenWithDeviceCodeAsync");
+            PublicClientApplication pca = new PublicClientApplication(labResponse.AppId);
+            var result = await pca.AcquireTokenWithDeviceCodeAsync(Scopes, deviceCodeResult =>
+            {
+                RunAutomatedDeviceCodeFlow(deviceCodeResult, labResponse.User);
+
+                return Task.FromResult(0);
+            }).ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.IsTrue(!string.IsNullOrEmpty(result.AccessToken));
+        }
+
+        private void RunAutomatedDeviceCodeFlow(DeviceCodeResult deviceCodeResult, LabUser user)
+        {
+            IWebDriver driver = null;
+            try
+            {
+                driver = InitDriver();
+
+                Debug.WriteLine("Navigating and enterign the code");
+
+                driver.Navigate().GoToUrl(deviceCodeResult.VerificationUrl);
+                driver.FindElement(By.Id("code")).SendKeys(deviceCodeResult.UserCode);
+
+                IWebElement continueBtn = WaitForElementToBeEnabled(driver, By.Id("continueBtn"));
+                continueBtn?.Click();
+
+                Debug.WriteLine("Logging in");
+                PerformLogin(driver, user);
+
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine("Browser automation failed " + ex);
+#if DESKTOP // Can't attach a file on netcore because mstest doesn't support it
+                Screenshot ss = ((ITakesScreenshot)driver).GetScreenshot();
+                string failurePicturePath = Path.Combine(TestContext.ResultsDirectory, "failure.png");
+                ss.SaveAsFile(failurePicturePath, ScreenshotImageFormat.Png);
+                TestContext.AddResultFile(failurePicturePath);
+
+                Debug.WriteLine("Failing because of " + ex);
+#endif
+                 throw;
+            }
+            finally
+            {
+                driver?.Close();
+                driver?.Dispose();
+            }
+        }
+
+        private static void PerformLogin(IWebDriver driver, LabUser user)
+        {
+            driver.FindElement(By.Id("i0116")).SendKeys(user.Upn); // username
+            driver.FindElement(By.Id("idSIButton9")).Click(); //Next
+            WaitForElementToBeEnabled(driver, By.Id("i0118")).SendKeys(user.Password); // password
+            WaitForElementToBeEnabled(driver, By.Id("idSIButton9")).Click(); // Finish
+        }
+
+        private static ChromeDriver InitDriver()
+        {
+            var chromeOptions = new ChromeOptions();
+            chromeOptions.AddArguments("headless"); // ~2x faster, no visual rendering
```

Comment:
This is the magic of headless browsers. I'm hopefully this will run on non-interactive windows sessions. When debugging this can be commented out.

Reply:
perfect. Perhaps add the "when debugging" comment above?

---
In reply to: [239847051](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/713#discussion_r239847051) [](ancestors = 239847051)